sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs

```diff
@@ -474,3 +476,264 @@ fn decode_leb128(idx: &mut usize, databytes: &[u8]) -> Result<u64> {
         "Corrupt data from HyperDebug GPIO monitoring".to_string(),
     ));
 }
+
+pub struct HyperdebugGpioBitbanging {
+    inner: Rc<Inner>,
+    cmsis_interface: BulkInterface,
+}
+
+impl HyperdebugGpioBitbanging {
+    /// CMSIS extension for HyperDebug GPIO.
+    const CMSIS_DAP_CUSTOM_COMMAND_GPIO: u8 = 0x83;
+
+    /// Sub-command for HyperDebug GPIO bitbanging.
+    const GPIO_BITBANG: u8 = 0x10;
+    const GPIO_BITBANG_STREAMING: u8 = 0x11;
+
+    pub fn open(inner: &Rc<Inner>, cmsis_interface: BulkInterface) -> Result<Self> {
+        // Exclusively claim CMSIS-DAP interface, preparing for bulk transfers.
+        inner
+            .usb_device
+            .borrow_mut()
+            .claim_interface(cmsis_interface.interface)?;
+        Ok(Self {
+            inner: Rc::clone(inner),
+            cmsis_interface,
+        })
+    }
+}
+
+impl GpioBitbanging for HyperdebugGpioBitbanging {
+    fn run(
+        &self,
+        pins: &[&dyn GpioPin],
+        clock_tick: Duration,
+        waveform: &mut [BitbangEntry],
+    ) -> Result<()> {
+        // Verify that `waveform` is valid, by converting into the binary representation to send
+        // to HyperDebug.
+        let mut encoded_waveform = encode_waveform(waveform, pins.len())?;
+
+        // Tell HyperDebug about the set of pins to manipulate, and the clock speed, using the
+        // textual console protocol.
+        let mut pin_names = Vec::new();
+        for pin in pins {
+            pin_names.push(
+                pin.get_internal_pin_name()
+                    .ok_or(TransportError::InvalidOperation)?,
+            );
+        }
+        self.inner.cmd_no_output(&format!(
+            "gpio bit-bang {} {}",
+            clock_tick.as_nanos(),
+            pin_names.join(" ")
+        ))?;
+
+        // Send an initial request, to ask how much buffer space HyperDebug has, so that we can
+        // fill the buffer, while avoiding overflows.
+        let usb = self.inner.usb_device.borrow();
+        let mut free_bytes: usize = {
+            let mut pkt = Vec::<u8>::new();
+            pkt.write_u8(Self::CMSIS_DAP_CUSTOM_COMMAND_GPIO)?;
+            pkt.write_u8(Self::GPIO_BITBANG)?;
+            pkt.write_u16::<LittleEndian>(0)?;
+            usb.write_bulk(self.cmsis_interface.out_endpoint, &pkt)?;
+
+            let mut databytes = [0u8; 64];
+
+            let c = usb.read_bulk(self.cmsis_interface.in_endpoint, &mut databytes)?;
+            let mut rdr = Cursor::new(&databytes[..c]);
+            ensure!(
+                rdr.read_u8()? == Self::CMSIS_DAP_CUSTOM_COMMAND_GPIO,
+                TransportError::CommunicationError(
+                    "Incorrect CMSIS-DAP header in response to GPIO request".to_string()
+                )
+            );
+            rdr.read_u8()?;
```

Comment:
What is the first byte of the header?

Reply:
This byte of the binary protocol is a status field (idle/active).  The textual command above should have already refused to set pins and clock period, in case there was ongoing bitbanging.  But for good measure, I have added an `ensure!` to verify that this byte also indicates "idle".